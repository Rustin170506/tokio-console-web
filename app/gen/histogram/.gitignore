# This file is intentionally left empty to ensure all files in this directory are tracked.
