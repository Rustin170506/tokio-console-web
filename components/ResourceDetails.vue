<template>
    <div>
        <div v-if="pending" class="flex items-center space-x-4">
            <USkeleton class="h-12 w-12" :ui="{ rounded: 'rounded-full' }" />
            <div class="space-y-2">
                <USkeleton class="h-4 w-[250px]" />
                <USkeleton class="h-4 w-[200px]" />
            </div>
        </div>
        <div v-else>
            <UCard
                :ui="{
                    body: {
                        base: 'xl:flex',
                    },
                }"
            >
                <div class="space-y-2 flex-1">
                    <UDivider type="dotted">
                        <div
                            class="flex items-center text-black dark:text-white font-bold text-md"
                        >
                            <span>Resource</span>
                        </div>
                    </UDivider>
                    <InfoField name="ID" :value="resourceBasicInfo.id" />
                    <InfoField
                        name="Parent ID"
                        :value="resourceBasicInfo.parent"
                    />
                    <InfoField name="Kind" :value="resourceBasicInfo.kind" />
                    <InfoField
                        name="Target"
                        :value="resourceBasicInfo.target"
                    />
                    <InfoField name="Type" :value="resourceBasicInfo.type" />
                    <InfoField name="Visibility">
                        <div class="flex justify-center items-center">
                            <UIcon
                                :name="resourceBasicInfo.visibilityIcon"
                                dynamic
                            />
                        </div>
                    </InfoField>
                    <InfoField
                        name="Location"
                        :value="resourceBasicInfo.location"
                    />
                </div>
                <UDivider
                    class="w-px mx-2"
                    color="gray"
                    orientation="vertical"
                />
                <div class="space-y-2 xl:ml-4 flex-1">
                    <UDivider
                        label="Attributes"
                        type="dotted"
                        :ui="{
                            label: 'text-black dark:text-white font-bold text-md',
                        }"
                    />
                    <div
                        v-for="(item, index) in resourceBasicInfo.attributes"
                        :key="index"
                        class="mb-1 w-10"
                    >
                        <p>
                            <span :class="item.name.class">{{
                                item.name.value
                            }}</span>
                            <span class="text-gray-500 dark:text-gray-400"
                                >=</span
                            >
                            <span :class="item.value.class">{{
                                item.value.value
                            }}</span>
                            <span :class="item.unit.class">{{
                                item.unit.value
                            }}</span>
                        </p>
                    </div>
                </div>
            </UCard>
            <AsyncOpsTable :resource-id="resourceId" />
        </div>
    </div>
</template>
<script setup lang="ts">
import { toResourceBasicInfo } from "~/types/resourceBasicInfo";

const router = useRouter();
const route = useRoute();
const resourceId = BigInt(route.params.id as string);
const { pending, resource, lastUpdatedAt } = useResourceDetails(resourceId);

if (resource === undefined || lastUpdatedAt.value === undefined) {
    router.push("/resources");
}

const resourceBasicInfo = computed(() => {
    return toResourceBasicInfo(resource!, lastUpdatedAt.value!);
});
</script>
