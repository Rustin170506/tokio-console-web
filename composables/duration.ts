export class Duration {
    seconds: bigint;
    nanos: number;

    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    subtract(other: Duration): Duration {
        const result = new Duration(
            this.seconds - other.seconds,
            this.nanos - other.nanos,
        );
        return result;
    }

    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result;
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }
}
