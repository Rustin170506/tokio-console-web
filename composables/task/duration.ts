// The Duration class represents a duration of time.
export class Duration {
    // The number of seconds in the duration.
    seconds: bigint;
    // The number of nanoseconds in the duration.
    nanos: number;

    // The constructor takes the number of seconds and nanoseconds.
    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
        // Normalize the duration so that nanos is less than 1e9 and non-negative.
        this.normalize();
    }

    // Normalize ensures that nanos is less than 1e9 and non-negative.
    normalize(): Duration {
        let adjustment = 0;
        // If nanos is 1e9 or more, we need to adjust seconds and nanos.
        if (this.nanos >= 1e9) {
            adjustment = Math.floor(this.nanos / 1e9);
            this.seconds += BigInt(adjustment);
            this.nanos -= adjustment * 1e9;
        } else if (this.nanos < 0) {
            // If nanos is negative, we need to adjust seconds and nanos.
            adjustment = Math.floor(-this.nanos / 1e9) + 1;
            this.seconds -= BigInt(adjustment);
            this.nanos += adjustment * 1e9;
        }
        return this;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    // subtract returns a new Duration that represents the difference between this duration and another.
    subtract(other: Duration): Duration {
        const seconds = this.seconds - other.seconds;
        const nanos = this.nanos - other.nanos;
        const result = new Duration(seconds, nanos);
        return result.normalize();
    }

    // add returns a new Duration that represents the sum of this duration and another.
    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result.normalize();
    }

    // greaterThan checks if this duration is greater than another.
    greaterThan(other: Duration): boolean {
        return (
            this.seconds > other.seconds ||
            (this.seconds === other.seconds && this.nanos > other.nanos)
        );
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }

    // valueOf returns the duration as a number of microseconds.
    // Used for sorting.
    valueOf(): number {
        return this.asMicroseconds();
    }

    // now returns the current time as a Duration from the Unix epoch.
    // TODO: use more precise time source. For example, performance.now() in browsers.
    static now(): Duration {
        return new Duration(BigInt(Math.floor(Date.now() / 1000)), 0);
    }

    static fromNano(nanos: bigint): Duration {
        const seconds = nanos / BigInt(1e9);
        const remainderNanos = nanos % BigInt(1e9);
        // After normalization, we can safely cast to number.
        return new Duration(seconds, Number(remainderNanos));
    }
}

// The Duration class can also be used as a timestamp.
export { Duration as Timestamp };
