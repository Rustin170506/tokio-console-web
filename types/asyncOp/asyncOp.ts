import type { Duration, Timestamp } from "../common/duration";
import type { AsyncOpStats } from "./asyncOpStats";

export class AsyncOp {
    id: bigint;
    parentIdStr: string;
    resourceId: bigint;
    metaId: bigint;
    source: string;
    stats: AsyncOpStats;

    constructor(
        id: bigint,
        parentIdStr: string,
        resourceId: bigint,
        metaId: bigint,
        source: string,
        stats: AsyncOpStats,
    ) {
        this.id = id;
        this.parentIdStr = parentIdStr;
        this.resourceId = resourceId;
        this.metaId = metaId;
        this.source = source;
        this.stats = stats;
    }

    totalDuration(since: Timestamp): Duration {
        return this.stats.total || since.subtract(this.stats.createdAt);
    }

    busyDuration(since: Timestamp): Duration {
        if (this.stats.lastPollStarted && !this.stats.lastPollEnded) {
            const currentTimeInPoll = since.subtract(
                this.stats.lastPollStarted,
            );
            return this.stats.busy.add(currentTimeInPoll);
        }
        return this.stats.busy;
    }

    idleDuration(since: Timestamp): Duration {
        return (
            this.stats.idle ||
            this.totalDuration(since).subtract(this.busyDuration(since))
        );
    }

    isDropped(): boolean {
        return !this.stats.total;
    }
}
