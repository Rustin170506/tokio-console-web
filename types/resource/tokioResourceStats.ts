import type { Duration, Timestamp } from "../common/duration";

export interface TokioResourceStats {
    createdAt: Timestamp;
    droppedAt?: Timestamp;
    total?: Duration;
    formattedAttributes: Array<Array<string>>;
}
