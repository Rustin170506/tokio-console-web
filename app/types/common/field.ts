import type { Metadata } from "./metadata";
import { Field as ProtoField } from "~/gen/common_pb";
import type { Location } from "~/gen/common_pb";

export function truncateRegistryPath(s: string): string {
    const regex = /.*\/cargo(\/registry\/src\/[^/]*\/|\/git\/checkouts\/)/;
    return s.replace(regex, "<cargo>/");
}

export function formatLocation(loc?: Location): string {
    if (loc) {
        let result = "";
        if (loc.modulePath) {
            result = loc.modulePath;
        } else if (loc.file) {
            const truncated = truncateRegistryPath(loc.file);
            result = truncated;
        } else {
            return "<unknown location>";
        }

        if (loc.line !== undefined) {
            result += `:${loc.line}`;

            if (loc.column !== undefined) {
                result += `:${loc.column}`;
            }
        }

        return result;
    }
    return "<unknown location>";
}

export enum FieldValueType {
    Bool = "Bool",
    Str = "Str",
    U64 = "U64",
    I64 = "I64",
    Debug = "Debug",
}

export class FieldValue {
    type: FieldValueType;
    value: boolean | string | bigint;

    constructor(type: FieldValueType, value: boolean | string | bigint) {
        this.type = type;
        this.value = value;
    }

    truncateRegistryPath(): FieldValue {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            typeof this.value === "string"
        ) {
            this.value = truncateRegistryPath(this.value);
        }
        return this;
    }

    ensureNonempty(): FieldValue | undefined {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            this.value === ""
        ) {
            return undefined;
        }
        return this;
    }

    static fromProto(field: ProtoField): FieldValue {
        switch (field.value.case) {
            case "debugVal":
                return new FieldValue(FieldValueType.Debug, field.value.value);
            case "strVal":
                return new FieldValue(FieldValueType.Str, field.value.value);
            case "u64Val":
                return new FieldValue(FieldValueType.U64, field.value.value);
            case "i64Val":
                return new FieldValue(FieldValueType.I64, field.value.value);
            case "boolVal":
                return new FieldValue(FieldValueType.Bool, field.value.value);
            default:
                throw new Error("Invalid field value case");
        }
    }
}

export class Field {
    static SPAWN_LOCATION = "spawn.location";
    static KIND = "kind";
    static NAME = "task.name";
    static TASK_ID = "task.id";

    name: string;
    value: FieldValue;

    constructor(name: string, value: FieldValue) {
        this.name = name;
        this.value = value;
    }

    static fromProto(field: ProtoField, meta: Metadata): Field | undefined {
        const nameValue = field.name;
        if (nameValue.case === undefined) {
            return undefined;
        }

        let name = "";
        switch (nameValue.case) {
            case "strName": {
                name = nameValue.value;
                break;
            }
            case "nameIdx": {
                const index = nameValue.value;
                const metaId = field.metadataId?.id;
                if (meta.id !== metaId) {
                    return undefined;
                }
                if (meta.fieldNames[Number(index)]) {
                    name = meta.fieldNames[Number(index)];
                } else {
                    return undefined;
                }
                break;
            }
        }

        let value = FieldValue.fromProto(field).ensureNonempty();
        // If the field is invalid or it has a string value which is empty, return null
        if (!name || !value) {
            return undefined;
        }

        // If the field name is 'SPAWN_LOCATION', truncate the registry path
        if (name === Field.SPAWN_LOCATION) {
            value = value.truncateRegistryPath();
        }

        return new Field(name.toString(), value);
    }
}
