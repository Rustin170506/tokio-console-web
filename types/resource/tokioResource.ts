/* eslint-disable camelcase */
import type { Duration, Timestamp } from "../common/duration";
import type { TokioResourceStats } from "./tokioResourceStats";
import { Resource_Kind, Resource_Kind_Known } from "~/gen/resources_pb";

export enum Visibility {
    Public,
    Internal,
}

export class TokioResource {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    parent: string;
    parentId: string;
    metaId: bigint;
    kind: string;
    stats: TokioResourceStats;
    target: string;
    concreteType: string;
    location: string;
    visibility: Visibility;

    constructor(
        id: bigint,
        spanId: bigint,
        parent: string,
        parentId: string,
        metaId: bigint,
        kind: string,
        stats: TokioResourceStats,
        target: string,
        concreteType: string,
        location: string,
        visibility: Visibility,
    ) {
        this.id = id;
        this.spanId = spanId;
        this.parent = parent;
        this.parentId = parentId;
        this.metaId = metaId;
        this.kind = kind;
        this.stats = stats;
        this.target = target;
        this.concreteType = concreteType;
        this.location = location;
        this.visibility = visibility;
    }

    totalDuration(since: Timestamp): Duration {
        const duration = since.subtract(this.stats.createdAt);
        return this.stats.total ?? duration;
    }

    isDropped(): boolean {
        return this.stats.total !== undefined;
    }
}

export function kindFromProto(pb: Resource_Kind): string {
    if (!pb.kind) {
        throw new Error("a resource should have a kind field");
    }

    switch (pb.kind.case) {
        case "known":
            if (pb.kind.value === Resource_Kind_Known.TIMER) {
                return "Timer";
            } else {
                throw new Error(`failed to parse known kind from ${pb.kind}`);
            }
        case "other":
            return pb.kind.value;
        default:
            throw new Error(`Unknown kind: ${pb.kind}`);
    }
}
