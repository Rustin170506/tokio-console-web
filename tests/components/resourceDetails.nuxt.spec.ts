import { expect, it } from "vitest";
import { mockNuxtImport, mountSuspended } from "@nuxt/test-utils/runtime";
import ResourceDetails from "~/components/ResourceDetails.vue";
import { Duration, Timestamp } from "~/types/common/duration";
import { TokioResource, Visibility } from "~/types/resource/tokioResource";
import { fromProtoResourceStats } from "~/types/resource/tokioResourceStats";
import { Stats as ProtoResourceStats } from "~/gen/resources_pb";
import type { Metadata } from "~/types/common/metadata";
import { AsyncOp } from "~/types/asyncOp/asyncOp";
import type { AsyncOpStats } from "~/types/asyncOp/asyncOpStats";
import { Field, FieldValue, FieldValueType } from "~/types/common/field";

mockNuxtImport("useRoute", () => {
    return () => {
        return {
            params: {
                id: "1",
            },
        };
    };
});

mockNuxtImport("useResourceDetails", () => {
    return (_id: bigint) => {
        const pending = ref<boolean>(false);
        const protoStats = new ProtoResourceStats({
            createdAt: { seconds: 1000n, nanos: 500 },
            droppedAt: { seconds: 2000n, nanos: 500 },
            attributes: [
                {
                    field: {
                        name: { case: "strName", value: "name" },
                        value: { case: "strVal", value: "test" },
                        metadataId: { id: 1n },
                    },
                    unit: "10ms",
                },
            ],
        });
        const metadata: Metadata = {
            id: 1n,
            target: "tokio::task",
            fieldNames: ["name"],
        };
        const stats = fromProtoResourceStats(protoStats, metadata);

        const resource = new TokioResource(
            1n,
            1n,
            "N/A",
            "N/A",
            1n,
            "Timer",
            stats,
            "tokio::time::driver::sleep",
            "Sleep",
            "console-subscriber/examples/grpc_web.rs:115:5",
            Visibility.Public,
        );
        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));

        return { pending, resource, lastUpdatedAt };
    };
});

mockNuxtImport("useAsyncOps", () => {
    return () => {
        const pending = ref<boolean>(false);
        const asyncOpsData = ref<Map<bigint, AsyncOp>>(new Map());
        const stats: AsyncOpStats = {
            createdAt: new Timestamp(1000n, 0),
            droppedAt: new Timestamp(2000n, 0),
            total: new Duration(1000n, 0),
            busy: new Duration(500n, 0),
            idle: new Duration(500n, 0),
            lastPollStarted: new Timestamp(1500n, 0),
            lastPollEnded: undefined,
            taskId: 1n,
            taskIdStr: "1",
            attributes: [
                {
                    field: new Field(
                        "name",
                        new FieldValue(FieldValueType.Str, "test"),
                    ),
                    unit: "10ms",
                },
            ],
            polls: 10n,
        };
        asyncOpsData.value.set(
            1n,
            new AsyncOp(1n, "parent", 1n, 1n, "source", stats),
        );

        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));
        return { pending, asyncOpsData, lastUpdatedAt };
    };
});

it("ResourceDetails", async () => {
    const component = await mountSuspended(ResourceDetails);
    expect(component.html()).toMatchSnapshot();
});
