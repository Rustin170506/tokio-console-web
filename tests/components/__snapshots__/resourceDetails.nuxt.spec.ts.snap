// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceDetails 1`] = `
"<div>
  <div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <!--v-if-->
      <div class="xl:flex px-4 py-5 sm:p-6">
        <div class="space-y-2 flex-1 mr-4">
          <div class="flex items-center align-center text-center w-full flex-row">
            <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
            <div class="font-medium text-gray-700 dark:text-gray-200 flex mx-3 whitespace-nowrap">
              <div class="flex items-center text-black dark:text-white font-bold text-md"><span>Resource</span></div>
            </div>
            <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
          </div>
          <div class="flex"><span class="font-bold text-black dark:text-white">ID</span><span class="mr-1">: </span><span class="">1</span></div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Parent ID</span><span class="mr-1">: </span><span class="">N/A</span></div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Kind</span><span class="mr-1">: </span><span class="">Timer</span></div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Target</span><span class="mr-1">: </span><span class="">tokio::time::driver::sleep</span></div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Type</span><span class="mr-1">: </span><span class="">Sleep</span></div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Visibility</span><span class="mr-1">: </span>
            <div class="flex justify-center items-center"><svg data-v-b3e50527="" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" role="img" class="icon" width="1em" height="1em" viewBox="0 0 24 24">
                <g fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5">
                  <path d="M2.036 12.322a1.012 1.012 0 0 1 0-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178c.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178"></path>
                  <path d="M15 12a3 3 0 1 1-6 0a3 3 0 0 1 6 0"></path>
                </g>
              </svg></div>
          </div>
          <div class="flex"><span class="font-bold text-black dark:text-white">Location</span><span class="mr-1">: </span><span class="">console-subscriber/examples/grpc_web.rs:115:5</span></div>
        </div>
        <div class="flex items-center align-center text-center flex-col w-px mx-2" color="gray">
          <div class="flex border-gray-200 dark:border-gray-800 h-full border-s border-solid"></div>
          <!--v-if-->
        </div>
        <div class="space-y-2 xl:ml-4 flex-1">
          <div class="flex items-center align-center text-center w-full flex-row">
            <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
            <div class="font-medium text-gray-700 dark:text-gray-200 flex mx-3 whitespace-nowrap"><span class="text-black dark:text-white font-bold text-md">Attributes</span></div>
            <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
          </div>
          <div class="mb-1 w-10">
            <p><span class="text-blue-500 font-bold dark:text-blue-300">name</span><span class="text-gray-500 dark:text-gray-400">=</span><span class="text-yellow-500 dark:text-yellow-300">test</span><span class="text-blue-500 dark:text-blue-300">10ms</span></p>
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900 mt-4">
      <!--v-if-->
      <div class="px-4 py-5 sm:p-6">
        <div class="flex items-center align-center text-center w-full flex-row">
          <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
          <div class="font-medium text-gray-700 dark:text-gray-200 flex mx-3 whitespace-nowrap"><span class="text-black dark:text-white font-bold text-md">Async Ops</span></div>
          <div class="flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted"></div>
        </div>
        <div class="relative overflow-x-auto">
          <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
            <thead class="relative">
              <tr class="">
                <!--v-if-->
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">ID</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Parent</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Task</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Source</span></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Total</span><span class="i-heroicons-bars-arrow-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Busy</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Idle</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
                    <!--v-if--><span class="">Polls</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
                  </button></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Attributes</span></th>
              </tr>
              <!--v-if-->
            </thead>
            <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
              <tr>
                <td colspan="9">
                  <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
                    <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
